mkosi/installer/apk.py

```diff
@@ -0,0 +1,194 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+import dataclasses
+from collections.abc import Sequence
+from pathlib import Path
+from typing import Optional
+
+from mkosi.config import Config
+from mkosi.context import Context
+from mkosi.installer import PackageManager
+from mkosi.log import die
+from mkosi.run import CompletedProcess, run, workdir
+from mkosi.util import _FILE, PathString
+
+
+@dataclasses.dataclass(frozen=True)
+class ApkRepository:
+    id: str
+    base_url: str
+    repo_type: str  # "alpine" or "postmarketos"
+    release: str
+    repo_name: Optional[str] = None
+
+    def __post_init__(self) -> None:
+        if self.release != "edge":
+            die(f"Only 'edge' release is currently supported, got '{self.release}'")
+
+    @property
+    def url(self) -> str:
+        if self.repo_type == "alpine":
+            return f"{self.base_url}/{self.release}/{self.repo_name}"
+        elif self.repo_type == "postmarketos":
+            release_path = f"v{self.release}"
+            if self.release == "edge":
+                release_path = "master"
+            if self.repo_name is None:
+                return f"{self.base_url}/{release_path}"
+            else:
+                return f"{self.base_url}/{self.repo_name}/{release_path}"
+        else:
+            raise ValueError(f"Unknown repo_type: {self.repo_type}")
+
+
+class Apk(PackageManager):
+    @classmethod
+    def executable(cls, config: Config) -> str:
+        return "apk"
+
+    @classmethod
+    def subdir(cls, config: Config) -> Path:
+        return Path("apk")
+
+    @classmethod
+    def package_subdirs(cls, cache: Path) -> list[tuple[Path, Path]]:
+        return [(Path("apk"), Path("apk"))]
+
+    @classmethod
+    def scripts(cls, context: Context) -> dict[str, list[PathString]]:
+        return {
+            "apk": cls.apivfs_script_cmd(context) + cls.cmd(context),
+            "mkosi-install":   ["apk", "add"],
+            "mkosi-upgrade":   ["apk", "upgrade"],
+            "mkosi-remove":    ["apk", "--remove", "del"],
+            "mkosi-reinstall": ["apk", "fix", "--reinstall"],
+        }  # fmt: skip
+
+    @classmethod
+    def mounts(cls, context: Context) -> list[PathString]:
+        mounts: list[PathString] = [
+            *super().mounts(context),
+        ]
+
+        cache_dir = context.config.package_cache_dir_or_default() / "apk"
+        cache_dir.mkdir(parents=True, exist_ok=True)
+        mounts += ["--bind", cache_dir, "/var/cache/apk"]
+
+        return mounts
+
+    @classmethod
+    def setup(cls, context: Context, repositories: Sequence[ApkRepository]) -> None:
+        config = context.root / "etc/apk/repositories"
```

Comment:
I would prefer we look at /etc/apk/repositories in the sandbox tree. Again, `--repositories-file` is looked up outside of the given root directory if the path is absolute so if we specify that to /etc/apk/repositories we can look up outside of the given root directory in the sandbox tree.

The only annoyance is that apk won't look at /etc/apk/repositories.d in that case. For know this means users won't be able to add additional repositories without overriding the default ones. You can make a feature request for apk to allow specifying a `.d` directory as well to fix this.